booster/library/Booster/SMT/Runner.hs

```diff
@@ -41,64 +46,108 @@ import Booster.Log (LoggerMIO (..), logMessage)
+    SMTOptions
+        { transcript = Nothing
+        , timeout = 125
+        , retryLimit = Just 3
+        , tactic = Nothing
+        }
+
 data SMTContext = SMTContext
-    { solver :: Backend.Solver
+    { options :: SMTOptions
+    , solver :: Backend.Solver
     , solverClose :: IO ()
-    , mbTranscript :: Maybe Handle
+    , mbTranscriptHandle :: Maybe Handle
+    , prelude :: [DeclareCommand]
     }
 
 ----------------------------------------
 {- TODO (later)
-- store prelude of [DeclareCommand] in context (enables hard resets)
 - error handling and retries
   - retry counter in context
-    - Reader becomes State
 - (possibly) run `get-info` on Unknown responses and enhance Unknown constructor
   - smtlib2: reason-unknown = memout | incomplete | SExpr
 -}
 
 mkContext ::
     LoggerMIO io =>
-    Maybe FilePath ->
+    SMTOptions ->
+    [DeclareCommand] ->
     io SMTContext
-mkContext transcriptPath = do
-    logMessage ("Starting new SMT solver" :: Text)
-    mbTranscript <-
-        forM transcriptPath $ \path -> do
-            logMessage $ "Transcript in file " <> pack path
-            liftIO $ do
-                h <- openFile path AppendMode
-                hSetBuffering h (BlockBuffering Nothing)
-                hSetBinaryMode h True
-                BS.hPutStrLn h "; starting solver process"
-                pure h
-    let config = Backend.defaultConfig
-    handle <- liftIO $ Backend.new config
-    solver <- liftIO $ Backend.initSolver Backend.Queuing $ Backend.toBackend handle
-    whenJust mbTranscript $ \h ->
+mkContext opts prelude = do
+    logMessage ("Starting SMT solver" :: Text)
+    (solver, handle) <- connectToSolver
+    mbTranscriptHandle <- forM opts.transcript $ \path -> do
+        logMessage $ "Transcript in file " <> pack path
+        liftIO $ do
+            h <- openFile path AppendMode
+            hSetBuffering h (BlockBuffering Nothing)
+            hSetBinaryMode h True
+            BS.hPutStrLn h "; starting solver process"
+            pure h
+    whenJust mbTranscriptHandle $ \h ->
         liftIO $ BS.hPutStrLn h "; solver initialised\n;;;;;;;;;;;;;;;;;;;;;;;"
-    logMessage ("Solver ready to use" :: Text)
     pure
         SMTContext
             { solver
             , solverClose = Backend.close handle
-            , mbTranscript
+            , mbTranscriptHandle
+            , prelude
+            , options = opts
             }
 
+{- | Close the connection to the SMT solver process, but hold on to the other resources in @SMTContext@,
+such as the transcript handle. This function is used in 'Booster.SMT.Interface.hardResetSolver'
+to reset the solver connection while keeping the same transcript handle.
+-}
 closeContext :: LoggerMIO io => SMTContext -> io ()
 closeContext ctxt = do
     logMessage ("Stopping SMT solver" :: Text)
-    whenJust ctxt.mbTranscript $ \h -> liftIO $ do
-        BS.hPutStrLn h "; stopping solver\n;;;;;;;;;;;;;;;;;;;;;;;"
+    liftIO ctxt.solverClose
```

Comment:
Maybe still write something to the transcript here?

Reply:
Good point!